custom_test.go

```diff
@@ -0,0 +1,87 @@
+package msgpack_test
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// ------------------------------------------------------------------------
+
+func init() {
+	msgpack.Register(reflect.TypeOf(customTestSet{}), encodeCustomTestSet, decodeCustomTestSet)
+}
+
+type customTestSet map[int]struct{}
+
+func encodeCustomTestSet(e *msgpack.Encoder, v reflect.Value) error {
+	if v.IsNil() {
+		return e.EncodeNil()
+	}
+
+	set := v.Interface().(customTestSet)
+	slc := make([]int, 0, len(set))
+	for n, _ := range set {
+		slc = append(slc, n)
+	}
+	return e.Encode(slc)
+}
+
+func decodeCustomTestSet(d *msgpack.Decoder, v reflect.Value) error {
+	if code, err := d.PeekCode(); err != nil {
+		return err
+	} else if code == msgpack.NilCode {
+		_, err = d.DecodeInterface()
+		return err
+	}
+
+	sln, err := d.DecodeSliceLen()
```

Comment:
DecodeSliceLen should return `-1, nil` when slice is nil. So you can remove code above and replace it with:

```
	sln, err := d.DecodeSliceLen()
	if err != nil {
		return err
	}
	if sln == -1 {
		// Slice is nil
		return nil
	}
```

Does it work work for you?